_posts/2020-09-28-biased-locking-help.adoc

```diff
@@ -0,0 +1,88 @@
+---
+
+Last week we became aware that the OpenJDK team in Java 15 have disabled so called *biased locking* (https://openjdk.java.net/jeps/374[JEP 374]) in the Java virtual machine. This is change from previous years and could *potentially* have a negative impact on Java application's performance.
+
+Red Hat's own performance teams are currently running performance tests to see how it affects our Java middleware, but no amount of generic testing can reveal how this change affects real world usage.
+
+That is where you come in.
+
+We would like to get info from you on whether your application performance is affected by biased locking or not.
+
+To do so please try the following in your application performance tests:
+
+Run your Quarkus application tests/performance as you normally would with the following command line flags on Java 11 (jdk11u) and if feasible Java 15 (jdk15):
+
+*enabled*: -XX:+UseBiasedLocking -XX:BiasedLockingStartupDelay=0
+
+*disabled*: -XX:-UseBiasedLocking
+
+We would like to know the result of this no matter if you see a regression or not on the same Java Virtual Machine.
+
+If you are able then running it with the following scenarios would be of great help too:
+
+ . single-threaded
+ . thread count ~= hardware core count
+ . thread count ~= N * hardware core count where 8 < N < 16
+
+ The intent of these are to see how the level of concurrency affects the result.
+
+=== Reporting the result
+
+Please open a bug on with `[jep374]` in the title + your project. i.e. `[jep374] results for acme project crazy panda` containing the following information per run in the description:
+
+```
+jvm used: jdk11 or jdk15
+biased locking: enabled or disabled
+thread-count: 1..N (if you know)
+hardware-core count: N (if you know)
+```
+
+This will help us immensly. Thank you!
+
+== Background
+
+Below is some background context on biased locking - all optional reading - you don't need to understand the details to help us by running your performance tests and let us know if anything changes.
+
+=== What is biased locking?
+
+Biased locking lowers the cost of /uncontended/ synchronization.
+
+*Without biased locking*: a thread needs to set and clear a lock bit
+when it performs repeated synchronizations on the same object. It also needs to wait for those set/clear writes to be drained to local cache before proceeding to execute further memory operations.
+
+*With biased locking*: the first time a thread synchronizes on an object it does a bit more work to acquire synchronized ('bias' it to the thread). Subsequent syncrhonizations proceed via a simple read test with no need to  drain to cache.
+
+Where's the trade off? Well, if a biased lock is contended then
+there is more work to do to bias and unbias the lock. However, it is known that many synchronized operations are uncontended.
+
+Additionally, so called 'lock-free' code (i.e. code that uses atomics or `cmpxchg` instead of 'synchronized' operations) bypasses these concerns.
```

Comment:
This sentence is incorrect and should just be removed:

`Additionally, so called 'lock-free' code ... `

Reply:
```suggestion
```